playback/src/audio_backend/alsa.rs

```diff
@@ -228,54 +236,96 @@ impl Sink for AlsaSink {
-            );
-            self.period_buffer
-                .extend_from_slice(&data[processed_data..processed_data + data_to_buffer]);
-            processed_data += data_to_buffer;
-            if self.period_buffer.len() == self.period_buffer.capacity() {
-                self.write_buf()?;
-                self.period_buffer.clear();
-            }
-        }
-
+        self.process_bytes(data, 0)?;
         Ok(())
     }
 }
 
 impl AlsaSink {
     pub const NAME: &'static str = "alsa";
 
-    fn write_buf(&mut self) -> io::Result<()> {
-        let pcm = self.pcm.as_mut().ok_or_else(|| {
-            io::Error::new(
-                io::ErrorKind::Other,
-                "Error writing from AlsaSink buffer to PCM, PCM is None",
-            )
-        })?;
-        let io = pcm.io_bytes();
-        if let Err(err) = io.writei(&self.period_buffer) {
-            // Capture and log the original error as a warning, and then try to recover.
-            // If recovery fails then forward that error back to player.
-            warn!(
-                "Error writing from AlsaSink buffer to PCM, trying to recover {}",
-                err
-            );
-            pcm.try_recover(err, false).map_err(|e| {
-                io::Error::new(
-                    io::ErrorKind::Other,
-                    format!(
-                        "Error writing from AlsaSink buffer to PCM, recovery failed {}",
-                        e
-                    ),
-                )
-            })?
+    fn configure_period_buffer(&mut self, bytes_per_period: usize) {
+        // Vec growth strategies can and have changed.
+        // Capacity may not always be trustworthy in the future.
+        // Use bytes_per_period to make sure that no matter what
+        // we give the PCM a period's worth of audio.
+        // Don't depend on len == capacity.
+        self.bytes_per_period = bytes_per_period;
+        // As noted above Vec growth strategies are not guaranteed
+        // so we over allocate initially so that our Vec will never
+        // actually reach full capacity thus (hopefully) avoiding
+        // future reallocations. Nearest (truncated) KiB + 2 KiB.
+        let over_allocation = ((bytes_per_period / KIB) * KIB) + TWO_KIB;
+        // If the capacity is greater than we want try to shrink it
+        // to it's current len (which should be zero) before trying
+        // to set the capacity with reserve_exact.
+        if self.period_buffer.capacity() > over_allocation {
+            self.period_buffer.shrink_to_fit();
+        }
+        // This does nothing if the capacity is already sufficient.
+        // Len should always be zero, but for the sake of being thorough...
+        self.period_buffer
+            .reserve_exact(over_allocation - self.period_buffer.len());
+    }
+
+    fn process_bytes(&mut self, data: &[u8], start_index: usize) -> io::Result<()> {
+        let space_left = self.bytes_per_period - self.period_buffer.len();
```

Comment:
You could then just use `self.period_buffer.capacity()` here. That also just references the Vec's "capacity field" so shouldn't be any slower, while saving the duplication.

Reply:
I know. You're the reason for most of these changes. Because you got it in your head that capacity could not be trusted.

Reply:
I yield. Sorry for having wasted your time.

Reply:
> I yield. Sorry for having wasted your time.

It's all good I just want to make sure you've made up you're mind which you seem to have.

Reply:
Done.